[SPIRV] Support non-constant indices for vector insert/extract

s-perron opened the pull request:
This patch updates the legalization of spv_insertelt and spv_extractelt to
handle non-constant (dynamic) indices. When a dynamic index is encountered, the
vector is spilled to the stack, and the element is accessed via OpAccessChain
(lowered from spv_gep).

This patch also adds custom legalization for G_STORE to scalarize vector stores
and refines the legalization rules for G_LOAD, G_STORE, and G_BUILD_VECTOR.

Fixes https://github.com/llvm/llvm-project/issues/170534

s-perron committed:
[SPIRV] Support non-constant indices for vector insert/extract

This patch updates the legalization of spv_insertelt and spv_extractelt to
handle non-constant (dynamic) indices. When a dynamic index is encountered, the
vector is spilled to the stack, and the element is accessed via OpAccessChain
(lowered from spv_gep).

This patch also adds custom legalization for G_STORE to scalarize vector stores
and refines the legalization rules for G_LOAD, G_STORE, and G_BUILD_VECTOR.

Fixes https://github.com/llvm/llvm-project/issues/170534

MrSidims reviewed (commented):
The approach is fine, though for future enhancements we may prevent spilling to the stack by using OpVectorInsertDynamic / OpVectorExtractDynamic, right?

s-perron committed:
Refactor to simplify code.

s-perron committed:
Fix G_FRAME_INDEX type assignment in spv_extractelt legalization

Restore the logic to assign a SPIR-V pointer-to-array type to the stack
temporary created during spv_extractelt legalization. This was accidentally
removed in the previous refactor and is required for legality.

s-perron committed:
Remove empty function.
